File: javascript/src/method/hasNot.ts
//··························································
// Copyright (c) 2023-2026. Jonathan Bédard ~ JóôòKiwi
//
// This project is free to use.
// All the right is reserved to the author of this project.
// My projects:
//  - https://github.com/joooKiwi/type
//  - https://github.com/joooKiwi/lazy
//  - https://github.com/joooKiwi/collection
//  - https://github.com/joooKiwi/enumeration
//··························································

import type {Nullable} from "@joookiwi/type"

import type {MinimalistCollectionHolder} from "../MinimalistCollectionHolder"
import type {CollectionHolder}           from "../CollectionHolder"

import {isArray}                       from "./isArray"
import {isArrayByStructure}            from "./isArrayByStructure"
import {isCollectionHolder}            from "./isCollectionHolder"
import {isCollectionHolderByStructure} from "./isCollectionHolderByStructure"
import {isMinimalistCollectionHolder}  from "./isMinimalistCollectionHolder"

//#region -------------------- Facade method --------------------

/**
 * Tell whenever the {@link value} <b>does not</b> exist in the {@link collection}
 *
 * @param collection The {@link Nullable nullable} collection ({@link MinimalistCollectionHolder}, {@link CollectionHolder} or {@link ReadonlyArray Array})
 * @param value      The value to compare
 * @return {boolean} <b>true</b> if the {@link value} is not present in the {@link collection}
 */
export function hasNot<const T, >(collection: Nullable<| MinimalistCollectionHolder<T> | CollectionHolder<T> | readonly T[]>, value: T,): boolean {
    if (collection == null)
        return true
    if (isCollectionHolder(collection,))
        return hasNotByCollectionHolder(collection, value,)
    if (isArray(collection,))
        return hasNotByArray(collection, value,)
    if (isMinimalistCollectionHolder(collection,))
        return hasNotByMinimalistCollectionHolder(collection, value,)

    if (isCollectionHolderByStructure<T>(collection,))
        return hasNotByCollectionHolder(collection, value,)
    if (isArrayByStructure<T>(collection,))
        return hasNotByArray(collection, value,)
    return hasNotByMinimalistCollectionHolder(collection, value,)
}


/**
 * Tell whenever the {@link value} <b>does not</b> exist in the {@link collection}
 *
 * @param collection The nullable collection
 * @param value      The value to compare
 * @return {boolean} <b>true</b> if the {@link value} is not present in the {@link collection}
 */
export function hasNotByMinimalistCollectionHolder<const T, >(collection: Nullable<MinimalistCollectionHolder<T>>, value: T,): boolean {
    if (collection == null)
        return true

    const size = collection.size
    if (size == 0)
        return true
    return __validate(collection, value, size,)
}

/**
 * Tell whenever the {@link value} <b>does not</b> exist in the {@link collection}
 *
 * @param collection The nullable collection
 * @param value      The value to compare
 * @return {boolean} <b>true</b> if the {@link value} is not present in the {@link collection}
 */
export function hasNotByCollectionHolder<const T, >(collection: Nullable<CollectionHolder<T>>, value: T,): boolean {
    if (collection == null)
        return true
    if (collection.isEmpty)
        return true
    return __validate(collection, value, collection.size,)
}

/**
 * Tell whenever the {@link value} <b>does not</b> exist in the {@link collection}
 *
 * @param collection The nullable collection
 * @param value      The value to compare
 * @return {boolean} <b>true</b> if the {@link value} is not present in the {@link collection}
 */
export function hasNotByArray<const T, >(collection: Nullable<readonly T[]>, value: T,): boolean {
    if (collection == null)
        return true

    const size = collection.length
    if (size == 0)
        return true
    return __validateByArray(collection, value, size,)
}

//#endregion -------------------- Facade method --------------------
//#region -------------------- Loop methods --------------------

function __validate<const T, >(collection: MinimalistCollectionHolder<T>, value: T, size: number,) {
    let index = -1
    while (++index < size)
        if (collection.get(index,) === value)
            return false
    return true
}

function __validateByArray<const T, >(collection: readonly T[], value: T, size: number,) {
    let index = -1
    while (++index < size)
        if (collection[index] === value)
            return false
    return true
}

//#endregion -------------------- Loop methods --------------------

File: java/collection.joookiwi/src/main/java/joookiwi/collection/java/method/HasNot.java
package joookiwi.collection.java.method;

import joookiwi.collection.java.CollectionHolder;
import joookiwi.collection.java.MinimalistCollectionHolder;
import joookiwi.collection.java.annotation.ExtensionFunction;
import joookiwi.collection.java.exception.ImpossibleConstructionException;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNullByDefault;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.annotations.Unmodifiable;

import static joookiwi.collection.java.CommonContracts.ALWAYS_FAIL_0;
import static joookiwi.collection.java.CommonContracts.IF_1ST_NULL_THEN_TRUE_2;

@NotNullByDefault
public final class HasNot
        extends Utility {

    @Contract(ALWAYS_FAIL_0)
    private HasNot() { throw new ImpossibleConstructionException("The utility class “Has” cannot be constructed.", HasNot.class); }

    //#region -------------------- Facade methods --------------------

    /// Tell whenever the `value` exist in the `collection`
    ///
    /// @param collection The [nullable][Nullable] [collection][MinimalistCollectionHolder]
    /// @param value      The value to compare
    /// @param <T>        The `collection` type
    /// @return `true` if the `value` is equals to no value in the `collection`
    @ExtensionFunction
    @Contract(IF_1ST_NULL_THEN_TRUE_2)
    public static <T extends @Nullable Object> boolean hasNot(final @Nullable MinimalistCollectionHolder<? extends T> collection,
                                                              final T value) {
        if (collection == null)
            return true;

        final var size = collection.size();
        if (size == 0)
            return true;
        return __validate(collection, size, value);
    }

    /// Tell whenever the `value` exist in the `collection`
    ///
    /// @param collection The [nullable][Nullable] [collection][CollectionHolder]
    /// @param value      The value to compare
    /// @param <T>        The `collection` type
    /// @return `true` if the `value` is equals to no value in the `collection`
    @ExtensionFunction
    @Contract(IF_1ST_NULL_THEN_TRUE_2)
    public static <T extends @Nullable Object> boolean hasNot(final @Nullable CollectionHolder<? extends T> collection,
                                                              final T value) {
        if (collection == null)
            return true;
        if (collection.isEmpty())
            return true;
        return __validate(collection, collection.size(), value);
    }

    /// Tell whenever the `value` exist in the `collection`
    ///
    /// @param collection The [nullable][Nullable] collection
    /// @param value      The value to compare
    /// @param <T>        The `collection` type
    /// @return `true` if the `value` is equals to no value in the `collection`
    @ExtensionFunction
    @Contract(IF_1ST_NULL_THEN_TRUE_2)
    public static <T extends @Nullable Object> boolean hasNot(final T @Nullable @Unmodifiable [] collection,
                                                              final T value) {
        if (collection == null)
            return true;

        final var size = collection.length;
        if (size == 0)
            return true;
        return __validate(collection, size, value);
    }

    //#endregion -------------------- Facade methods --------------------
    //#region -------------------- Loop methods --------------------

    private static <T extends @Nullable Object> boolean __validate(final MinimalistCollectionHolder<? extends T> collection,
                                                                   final int size,
                                                                   final T value) {
        var index = -1;
        if (value == null) {
            while (++index < size)
                if (collection.get(index) == null)
                    return false;
            return true;
        }

        while (++index < size) {
            final var collectionValue = collection.get(index);
            if (value == collectionValue || value.equals(collectionValue))
                return false;
        }
        return true;
    }

    private static <T extends @Nullable Object> boolean __validate(final T @Unmodifiable [] collection,
                                                                   final int size,
                                                                   final T value) {
        var index = -1;
        if (value == null) {
            while (++index < size)
                if (collection[index] == null)
                    return false;
            return true;
        }

        while (++index < size) {
            final var collectionValue = collection[index];
            if (value == collectionValue || value.equals(collectionValue))
                return false;
        }
        return true;
    }

    //#endregion -------------------- Loop methods --------------------

}
